source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h

```diff
@@ -210,64 +218,122 @@ class UdpProxyFilter : public Network::UdpListenerReadFilter,
+        hash = absl::HashOf(hash, value.host_.address()->asStringView());
+      }
+      return hash;
     }
     size_t operator()(const ActiveSession* value) const {
-      return absl::Hash<const Network::UdpRecvData::LocalPeerAddresses>()(value->addresses());
+      LocalPeerHostAddresses key{value->addresses(), value->host()};
+      return this->operator()(key);
     }
+    size_t operator()(const ActiveSessionPtr& value) const { return this->operator()(value.get()); }
+
+  private:
+    const bool consider_host_;
   };
 
   struct HeterogeneousActiveSessionEqual {
     // See description for HeterogeneousActiveSessionHash::is_transparent.
     using is_transparent = void; // NOLINT(readability-identifier-naming)
 
+    HeterogeneousActiveSessionEqual(const bool consider_host) : consider_host_(consider_host) {}
+
     bool operator()(const ActiveSessionPtr& lhs,
                     const Network::UdpRecvData::LocalPeerAddresses& rhs) const {
       return lhs->addresses() == rhs;
     }
-    bool operator()(const ActiveSessionPtr& lhs, const ActiveSessionPtr& rhs) const {
-      return lhs->addresses() == rhs->addresses();
+    bool operator()(const ActiveSessionPtr& lhs, const LocalPeerHostAddresses& rhs) const {
+      return this->operator()(lhs, rhs.local_peer_addresses_) &&
+             (consider_host_ ? &lhs->host() == &rhs.host_ : true);
     }
     bool operator()(const ActiveSessionPtr& lhs, const ActiveSession* rhs) const {
-      return lhs->addresses() == rhs->addresses();
+      LocalPeerHostAddresses key{rhs->addresses(), rhs->host()};
+      return this->operator()(lhs, key);
     }
+    bool operator()(const ActiveSessionPtr& lhs, const ActiveSessionPtr& rhs) const {
+      return this->operator()(lhs, rhs.get());
+    }
+
+  private:
+    const bool consider_host_;
   };
 
   /**
    * Wraps all cluster specific UDP processing including session tracking, stats, etc. In the future
    * we will very likely support different types of routing to multiple upstream clusters.
    */
   class ClusterInfo {
+  protected:
+    using SessionStorageType = absl::flat_hash_set<ActiveSessionPtr, HeterogeneousActiveSessionHash,
+                                                   HeterogeneousActiveSessionEqual>;
+
   public:
-    ClusterInfo(UdpProxyFilter& filter, Upstream::ThreadLocalCluster& cluster);
-    ~ClusterInfo();
-    Network::FilterStatus onData(Network::UdpRecvData& data);
+    ClusterInfo(UdpProxyFilter& filter, Upstream::ThreadLocalCluster& cluster,
+                SessionStorageType&& sessions);
+    virtual ~ClusterInfo();
+    virtual Network::FilterStatus onData(Network::UdpRecvData& data) PURE;
     void removeSession(const ActiveSession* session);
 
     UdpProxyFilter& filter_;
     Upstream::ThreadLocalCluster& cluster_;
     UdpProxyUpstreamStats cluster_stats_;
 
+  protected:
+    using HostConstSharedPtrOptConstRef =
+        absl::optional<std::reference_wrapper<const Upstream::HostConstSharedPtr>>;
```

Comment:
nit: can you just removed this and pass nullptr in cases where this is no host? I don't think we need to differentiate between null and not set at all?

Reply:
done